pkg/plugins/runtime/k8s/controllers/mesh_defaults_controller.go

```diff
@@ -19,9 +27,29 @@ type MeshDefaultsReconciler struct {
 }
 
 func (r *MeshDefaultsReconciler) Reconcile(req kube_ctrl.Request) (kube_ctrl.Result, error) {
+	mesh := core_mesh.NewMeshResource()
+	if err := r.ResourceManager.Get(context.Background(), mesh, store.GetByKey(req.Name, core_model.NoMesh)); err != nil {
+		return kube_ctrl.Result{}, errors.Wrap(err, "could not get default mesh resources")
+	}
+
+	// Before creating default policies for the mesh we want to ensure that this mesh wasn't processed before.
+	// We can't rely on filtering by CreateFunc, because apparently it sends the create event every time resource
+	// is added to the underlying Informer. That's why on the Kuma CP restart Mesh will be processed the second time
+	if processed := mesh.GetMeta().(*k8s.KubernetesMetaAdapter).GetAnnotations()[common_k8s.K8sProcessed]; processed == "true" {
+		return kube_ctrl.Result{}, nil
+	}
+
 	if err := defaults_mesh.EnsureDefaultMeshResources(r.ResourceManager, req.Name); err != nil {
 		return kube_ctrl.Result{}, errors.Wrap(err, "could not create default mesh resources")
 	}
```

Comment:
Looking at what `EnsureDefaultMeshResources` does, we can collapse all the `ensureDefaultXXX` helper functions into a single generic function. Can you file an issue to clean that up later?

Reply:
Done https://github.com/kumahq/kuma/issues/2540